MGMT-1173 Run assistant using podman

rwsu opened the pull request:
Adds ISOGenerator and InstallConfigGenerator interfaces to allow k8s and local implementations.

Moves k8s job specific code to job.go.

Introduces deploy-onprem, clean-onprem, and test-onprem make targets.

Dockerfile.bm-inventory-disconnected combines Dockerfile.bm-inventory with additional bits needed to bake install config and, in a follow-on patch, ISO generation scripts into the bm-inventory image.

eranco74 commented on Makefile:
```diff
@@ -175,3 +175,22 @@ subsystem-clean:
+	podman run -dt --pod assisted-installer --env-file onprem-environment -v s3-volume:/mnt/data:rw --name s3 scality/s3server:latest
+	podman run -dt --pod assisted-installer --env-file onprem-environment --name mariadb mariadb:latest
+	sleep 15
+	podman run -dt --pod assisted-installer --env-file onprem --name installer ${SERVICE}
```
did you mean?
```suggestion
	podman run -dt --pod assisted-installer --env-file onprem-environment --name installer ${SERVICE}
```

rwsu replied:
Yes. Thank you.

YuviGold commented on Dockerfile.bm-inventory-disconnected:
```diff
@@ -0,0 +1,27 @@
+
+LABEL "git_revision"=${GIT_REVISION}
+
+RUN dnf install -y python pip libvirt-devel
```
Try to unite dockerfile steps as much as possible to reduce layers and the image size
e.g. 
```
RUN dnf install -y python pip libvirt-devel && pip install boto3 botocore
```

eranco74 replied:
Cleaning up after dnf will help
```suggestion
RUN dnf install -y python pip libvirt-devel && dnf clean all
```
Don't think installing pip in the same layer will reduce image size

YuviGold replied:
@eranco74 Each layer counts - doesn't matter its content.

eranco74 replied:
@YuviGold that is just wrong.
```
$ cat dockerfile.1l
FROM fedora:31
RUN dnf install -y python pip libvirt-devel && pip install boto3
$ cat dockerfile.2l
FROM fedora:31
RUN dnf install -y python pip libvirt-devel
RUN pip install boto3
$ docker images | grep layer
2layer                                                        latest                                     eed2166e776e        8 minutes ago       554MB
1layer                                                        latest                                     7a902c214e64        10 minutes ago      554MB

```

YuviGold commented on Makefile:
```diff
@@ -175,3 +175,22 @@ subsystem-clean:
 clear-deployment:
 	-python3 ./tools/clear_deployment.py --delete-namespace $(APPLY_NAMESPACE) || true
+
+deploy-onprem:
```
bm-inventory already have a way to test onprem with minikube deployment
```bash
skipper make clear-all deploy-test test
```
Does this method has any advantages?

romfreiman replied:
@YuviGold it's inside podman, not minikube. This exactly the solution they are working on - to run assisted in non-k8s env

eranco74 commented on pkg/job/job.go:
```diff
@@ -23,26 +52,47 @@ type API interface {
 	RetryInterval       time.Duration `envconfig:"JOB_RETRY_INTERVAL" default:"1s"`
 	RetryAttempts       int           `envconfig:"JOB_RETRY_ATTEMPTS" default:"30"`
+	ImageBuilder        string        `envconfig:"IMAGE_BUILDER" default:"quay.io/ocpmetal/installer-image-build:latest"`
+	ImageBuilderCmd     string        `envconfig:"IMAGE_BUILDER_CMD" default:"echo hello"`
```
cmd seems off

rwsu replied:
That line was moved from inventory.go. It also seemed weird to me. Does anyone know why it is there? Is it for debugging purposes?

filanov replied:
yes it's for debug. i will do a cleanup on this.

filanov replied:
`ImageBuilderCmd` don't longer exists please remove

rwsu replied:
Removed ImageBuilderCmd.

eranco74 commented on cmd/main.go:
```diff
@@ -134,8 +115,36 @@ func main() {
+			kclient = nil
+		}
+
+		generator = job.New(log.WithField("pkg", "k8s-job-wrapper"), kclient, Options.JobConfig)
```
Nice!

eranco74 commented on internal/bminventory/inventory.go:
```diff
@@ -467,38 +377,19 @@ func (b *bareMetalInventory) GenerateClusterISO(ctx context.Context, params inst
 	}
-
-	// Kill the previous job in case it's still running
-	prevJobName := fmt.Sprintf("createimage-%s-%s", cluster.ID, previousCreatedAt.Format("20060102150405"))
```
This might be required, consider adding it to the generator interface

rwsu replied:
It has been moved to GenerateISO in pkg/job/job.go.

romfreiman commented on Dockerfile.bm-inventory-disconnected:
```diff
@@ -0,0 +1,27 @@
+LABEL "git_revision"=${GIT_REVISION}
+
+RUN dnf install -y python pip libvirt-devel
+RUN pip install boto3
```
same layer please

rwsu replied:
Ack. Moving it to the same layer.

eranco74 replied:
Moving to the same layer will not change the image size.
On the other chaning line 12 to:
```
RUN dnf install -y python pip libvirt-devel && dnf clean all
```
will save 222MB
Try it

tsorya commented on Makefile:
```diff
@@ -169,3 +169,21 @@ subsystem-clean:
 	-python3 ./tools/clear_deployment.py --delete-namespace $(APPLY_NAMESPACE) || true
+
+deploy-onprem:
+	podman pod create --name assisted-installer -p 3306,8000,8090,8080
```
i think u need to something like 
CONTAINER_COMMAND = $(shell if [ -x "$(shell command -v docker)" ];then echo "docker" ; else echo "podman";fi)
As for me it must deployed with podman and docker

rwsu replied:
Done.

rwsu commented:
Hi @filanov, please review. It also contains the ISOGenerator and InstallConfigGenerator interfaces discussed in https://github.com/rwsu/bm-inventory/pull/1.

filanov commented on cmd/main.go:
```diff
@@ -141,11 +122,39 @@ func main() {
+	var generator job.ISOInstallConfigGenerator
 
-	bm := bminventory.NewBareMetalInventory(db, log.WithField("pkg", "Inventory"), hostApi, clusterApi, Options.BMConfig, jobApi, eventsHandler, s3Client, metricsManager)
+	if Options.Target != "onprem" {
```
We already have `UseK8s` do we really need to have two flags? We can use `Target` but then `UseK8s` should be removed.

mhrivnak replied:
2 and 3 should eventually become "On local deployment which uses generation without S3". More details in the `S3` section here: https://github.com/openshift/enhancements/pull/411/files

ronniel1 replied:
That is fine.
The point is that anyway, we should not have both flags

filanov commented on internal/bminventory/inventory_test.go:
```diff
@@ -185,7 +179,7 @@ var _ = Describe("GetNextSteps", func() {
 		mockEvents = events.NewMockHandler(ctrl)
 		mockJob = job.NewMockAPI(ctrl)
-		mockJob.EXPECT().Create(gomock.Any(), gomock.Any()).Return(nil).Times(1)
+		mockJob.EXPECT().GenerateISO(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nil).Times(1)
```
if you have the same mock multiple times you can use unnamed functions to set it once then if the api will change you have less places to fix.

rwsu replied:
Done.

filanov commented on pkg/job/job.go:
```diff
@@ -2,19 +2,46 @@ package job
+
+// [TODO]  make sure that we use openshift-installer from the release image, otherwise the KubeconfigGenerator image must be updated here per openshift version
+
+type ISOGenerator interface {
```
I don't that this is related to the job package 
The generator can use job as a way to implement iso generation.

ronniel1 replied:
The generator interface has multiple implementations.
1 of then uses the job

rwsu replied:
I've moved the interfaces to a generator package.

ronniel1 commented on internal/bminventory/inventory.go:
```diff
@@ -504,38 +414,19 @@ func (b *bareMetalInventory) GenerateClusterISO(ctx context.Context, params inst
 			WithPayload(common.GenerateError(http.StatusInternalServerError, formatErr))
 	}
 
-	// This job name is exactly 63 characters which is the maximum for a job - be careful if modifying
```
This seems like an important comment. 
Hy was it removed?
Although I would rather have it as a validation in the code or UT than as a comment, but still

rwsu replied:
This comment along with some following lines have been moved to GenerateISO in job.go.

ronniel1 commented on onprem-environment:
```diff
@@ -0,0 +1,12 @@
+S3DATAPATH=/mnt/data
+S3METADATAPATH=/mnt/data
+S3_ENDPOINT_URL=http://127.0.0.1:8000
+MYSQL_DATABASE=installer
+MYSQL_PASSWORD=admin
+MYSQL_ROOT_PASSWORD=root
+MYSQL_USER=admin
+DB_HOST=127.0.0.1
```
These need to be updated due to the postgres change

rwsu replied:
Updated to postgres.

ronniel1 commented on pkg/job/job.go:
```diff
@@ -2,19 +2,46 @@ package job
+
+// [TODO] need to find more generic way to set the openshift release image
+//https://mirror.openshift.com/pub/openshift-v4/clients/ocp-dev-preview/4.5.0-0.nightly-2020-05-21-015458/
+const OverrideImageName = "quay.io/openshift-release-dev/ocp-release-nightly@sha256:a9f7564e0f2edef2c15cc1da699ebd1d11f5acd717c3668940848b3fed0d13c7"
```
We plan to change this.
It will come from an environment variable
https://issues.redhat.com/browse/MGMT-1616

rwsu commented on internal/bminventory/inventory.go:
```diff
@@ -164,111 +152,33 @@ func NewBareMetalInventory(
 }
 
-func (b *bareMetalInventory) generateDummyISOImage() {
+func (b *bareMetalInventory) generateDummyISOImage(generator job.ISOInstallConfigGenerator) {
```
Done.

rwsu commented on Dockerfile.bm-inventory-onprem:
```diff
@@ -0,0 +1,23 @@
+# [TODO] - remove this line, once we are ready to use openshift-installer from the release
+FROM quay.io/yshnaidm/openshift-installer AS openshift-installer
```
Changed. @ronniel1, does that TODO still apply?

ronniel1 replied:
I think it is.
@yevgeny-shnaidman ?

yevgeny-shnaidman replied:
yes, it still applies

rwsu commented on README.md:
```diff
@@ -192,6 +192,35 @@ skipper make deploy-all DEPLOY_TAG=<tag>
+This type of deployment requires a different installer image that combines components that are used to generate the installer ISO and configuration files. First build and push this image to a repository using Dockerfile.bm-inventory-onprem:
+
+`skipper make all`
+`podman build -f Dockerfile.bm-inventory-onprem -t quay.io/<your-org>/bm-inventory:latest . `
```
Done.

rwsu commented on README.md:
```diff
@@ -192,6 +192,35 @@ skipper make deploy-all DEPLOY_TAG=<tag>
+
+`skipper make all`
+`podman build -f Dockerfile.bm-inventory-onprem -t quay.io/<your-org>/bm-inventory:latest . `
+`podman push quay.io/<your-org>/bm-inventory:latest`
```
Done.

rwsu commented:
> What about tests?
> Does it make sense to UT some of this?
> What about subsystem?

I will add them shortly.

ronniel1 commented on README.md:
```diff
@@ -192,6 +192,36 @@ skipper make deploy-all DEPLOY_TAG=<tag>
+This type of deployment requires a different installer image that combines components that are used to generate the installer ISO and configuration files. First build and push this image to a repository using Dockerfile.bm-inventory-onprem:
+
+`skipper make all`
+`export SERVICE=quay.io/<your-org>/bm-inventory:latest`
```
This should be the first command

filanov commented on internal/bminventory/inventory.go:
```diff
@@ -204,93 +192,14 @@ func (b *bareMetalInventory) generateDummyISOImage() {
 	// create dummy job without uploading to s3, we just need to pull the image
-	if err := b.job.Create(requestid.ToContext(context.Background(), requestID),
-		b.createImageJob(jobName, imgName, "Dummy", false)); err != nil {
+	if err := b.generator.GenerateISO(requestid.ToContext(context.Background(), requestID), cluster, jobName, imgName, "Dummy", b.eventsHandler); err != nil {
```
If you take a look at the existing code you will see that `createImageJob` have another parameter `performUpload bool`
if it is `fales` we don't want to upload the iso to S3
It looks like this behavior will break with the new code. 
https://github.com/filanov/bm-inventory/pull/376/files#diff-97c2fe56dacc4fee9ab38e4dbd281ab8L223

rwsu replied:
I've updated job.GenerateISO to set performUpload correctly if ignitionConfig is "Dummy".
https://github.com/filanov/bm-inventory/pull/376/files#diff-b73c540b6ff88e46556b97a0c803a5d2R278

filanov commented on internal/bminventory/inventory_test.go:
```diff
@@ -79,7 +77,7 @@ var _ = Describe("GenerateClusterISO", func() {
 		mockEvents = events.NewMockHandler(ctrl)
 		mockJob = job.NewMockAPI(ctrl)
-		mockJob.EXPECT().Create(gomock.Any(), gomock.Any()).Return(nil).Times(1)
+		mockJob.EXPECT().GenerateISO(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nil).Times(1)
```
looks like this mock return itself many times maybe you can create an unnamed function to set it
then if the api will change you will need to change only one place
```
	mockGenerateIsoSuccess := func() {
		mockJob.EXPECT().GenerateISO(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nil).Times(1)
	}
```

rwsu replied:
Done. Thanks for the tip.

ohadlevy commented on Dockerfile.bm-inventory-onprem:
```diff
@@ -0,0 +1,36 @@
+
+LABEL "git_revision"=${GIT_REVISION}
+
+RUN dnf install -y python pip libvirt-devel && dnf clean all && pip install boto3 botocore
```
nitpicking, but why do we install devel deps on a "prod" container?

rwsu commented on cmd/main.go:
```diff
@@ -63,6 +64,7 @@ var Options struct {
 	ImageExpirationInterval     time.Duration `envconfig:"IMAGE_EXPIRATION_INTERVAL" default:"30m"`
 	ImageExpirationTime         time.Duration `envconfig:"IMAGE_EXPIRATION_TIME" default:"60m"`
 	ClusterConfig               cluster.Config
+	Target                      string `envconfig:"TARGET" default:""`
```
Removed USE_K8S and changed TARGET to DEPLOY_TARGET.

ronniel1 commented:
Running jenkins using the new stages fails: http://10.46.10.2:8080/job/bm-inventory/job/PR-535/2/console
Also notice that in inventory.go, function ResetCluster() new logic relating to the job was added which needs to be handled in this PR, I guess

rwsu commented:
> Running jenkins using the new stages fails: http://10.46.10.2:8080/job/bm-inventory/job/PR-535/2/console
> Also notice that in inventory.go, function ResetCluster() new logic relating to the job was added which needs to be handled in this PR, I guess

The Jenkins stage failed because the "build-onprem" target was missing. I have included it in the Makefile.

I also added the reset logic to this PR. I added an AbortInstallConfig interface to the generator.

pawanpinjarkar commented on Makefile:
```diff
@@ -134,6 +137,14 @@ deploy-test:
+	podman run -dt --pod assisted-installer --env-file onprem-environment -v s3-volume:/mnt/data:rw --name s3 scality/s3server:latest
+	podman run -dt --pod assisted-installer --env-file onprem-environment --name db postgres:12.3-alpine
+	podman run -dt --pod assisted-installer --env-file onprem-environment --restart always --name installer ${SERVICE}
+	podman run -dt --pod assisted-installer --env-file onprem-environment -v $(PWD)/deploy/ui/nginx.conf:/opt/bitnami/nginx/conf/server_blocks/nginx.conf:z --name ui quay.io/ocpmetal/ocp-metal-ui:latest
```
The default image pull policy for podman is `missing`. Can you change it to `always`?

podman run -dt --pod assisted-installer --env-file onprem-environment **--pull always** -v $(PWD)/deploy/ui/nginx.conf:/opt/bitnami/nginx/conf/server_blocks/nginx.conf:z --name ui quay.io/ocpmetal/ocp-metal-ui:latest

rwsu replied:
I understand why you want "--pull always" to give you the latest UI when developing/testing. But when this is deployed using an ISO, the images will be pre-baked so asking it to always pull will fail there. I can add it for now, but we will need to remember to disable --pull always from the final delivery mechanism.

rwsu commented:
It is unclear to me what the lint error in the CI is about. Running "make lint" locally on this branch passes.
